BlueDBOnDisk/src/main/java/org/bluedb/disk/backup/BackupManager.java

```diff
@@ -58,39 +82,124 @@ private void copyMetaData(ReadWriteCollectionOnDisk<?> collection, Path tempFold
 		FileUtils.copyDirectoryWithoutLock(srcPath, dstPath);
 	}
 
-	private void copyDataFolders(ReadWriteCollectionOnDisk<?> collection, Path tempFolder) throws BlueDbException {
+	private void copyDataFolders(ReadWriteCollectionOnDisk<?> collection, Path tempFolder, Range includedTimeRange) throws BlueDbException {
+		Range fileRangeToConsider = createFileRangeToConsider(includedTimeRange, collection.isTimeBased());
+		
 		for (ReadWriteSegment<?> segment: collection.getSegmentManager().getAllExistingSegments()) {
-			copyDataFolders(segment, tempFolder);
+			copyDataFolders(segment, tempFolder, includedTimeRange, collection.isTimeBased(), fileRangeToConsider);
+		}
+	} 
+
+	private Range createFileRangeToConsider(Range includedTimeRange, boolean isTimeBased) {
+		if(!isTimeBased) {
+			return Range.createMaxRange(); //Include everything in a non time base collection
 		}
+	
+		/*
+		 * Since values with a TimeFrameKey have an end time, we have to look at anything with a start time at or
+		 * before the end time of the included time range since any of those objects could be long enough to overlap
+		 * into the included time range. And since you can insert a value with a TimeFrameKey into a Collection with
+		 * a key type of TimeKey we must do this even for collections with a key type of TimeKey. 
+		 */
+		return new Range(Long.MIN_VALUE, includedTimeRange.getEnd());
 	}
 
-	private void copyDataFolders(ReadWriteSegment<?> segment, Path tempFolder) throws BlueDbException {
-		Range range = new Range(Long.MIN_VALUE, Long.MAX_VALUE);
-		List<File> files = segment.getOrderedFilesInRange(range);
+	private void copyDataFolders(ReadWriteSegment<?> segment, Path tempFolder, Range includedTimeRange, boolean isTimeBased, Range fileRangeToConsider) throws BlueDbException {
+		List<File> files = segment.getOrderedFilesInRange(fileRangeToConsider);
 		for (File file: files) {
 			Range fileRange = Range.fromUnderscoreDelmimitedString(file.getName());
 			long groupingNumber = fileRange.getStart();
-			try (BlueReadLock<Path> lock = segment.getReadLockFor(groupingNumber)) {
-				Path src = lock.getKey();
-				Path dst = translatePath(dbPath, tempFolder, src);
-				FileUtils.copyFileWithoutLock(src, dst);  // already have read lock on src, shouldn't need write lock on dst
+			
+			if(isTimeBased && !includedTimeRange.isMaxRange()) {
+				copyDataFileAfterFilteringBasedOnTime(segment, tempFolder, includedTimeRange, groupingNumber);
+			} else {
+				copyDataFileStraightOver(segment, tempFolder, groupingNumber);
 			}
 		}
 	}
 
-	private void copyChanges(ReadWriteCollectionOnDisk<?> collection, long backupStartTime, long backupEndTime, Path tempFolder) throws BlueDbException {
+	private void copyDataFileAfterFilteringBasedOnTime(ReadWriteSegment<?> segment, Path tempFolder, Range includedTimeRange, long groupingNumber) throws BlueDbException {
+		Path src = segment.getPathFor(groupingNumber);

```

Comment:
So in the method after this one "copyDataFileStriaghtOver" we create a lock on the path. We specifically don't do this here. What was the thinking there? Or rather when do and when do we not use locks when reading/deleting files?

Reply:
In general, we acquire a read lock for all file reads and a write lock for all file writes. The backup process is a bit of an exception to this rule. We need a lock on the files we're reading from because other read/write queries could be attempting to read or write to/from them during the backup process. However, the files we're writing to are only touched by this process. It creates a temp directory which it writes all of the backup files to before zipping it all up. There is no need to expend resources protecting those backup files from other threads accessing them. Though the resource expenditure is probably not very significant.

Notes:
Line 127 - We open an output stream without a lock for writing to the backup file
Line 128 - We open up an input stream for the file we will read from. This method acquires a read lock for us.
Line 145 - Instead of opening an input stream that allows us to read one object at a time, we just straight up acquire a read lock so that we can copy the bytes straight over.
Line 148 - This writes the bytes to the backup file without a lock